### PR TITLE
Adding more types

### PR DESCRIPTION
**Pull request by mamazu**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | yes
| Deprecations? | yes
| Fixed tickets | -
| Related issues/PRs | -
| License | MIT
| Documentation PR | -

#### What's in this PR?

* Improving typing
* Fixing a bug in the src/Sulu/Component/Rest/RequestParametersTrait.php

#### Why?

Improving the types to mirror the actual return and parameter types is always good.

#### To Do

- [x] Add breaking changes to UPGRADE.md

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php [hunk 1]
```diff
@@ -156,8 +156,8 @@ public function getAction($id, Request $request)
         try {
             $locale = $this->getRequestParameter($request, 'locale', true);
             $depth = \intval($request->get('depth', 0));
-            $breadcrumb = $this->getBooleanRequestParameter($request, 'breadcrumb', false, false);
-            $children = $this->getBooleanRequestParameter($request, 'children', false, false);
+            $breadcrumb = $this->getBooleanRequestParameter($request, 'breadcrumb', false) ?? false;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php]: this `??` ar enot required as false is the default value set and returned. phpdoc of that method can be improved so phpstan understands it.

File: src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php [hunk 2]
```diff
@@ -156,8 +156,8 @@ public function getAction($id, Request $request)
...
+            $children = $this->getBooleanRequestParameter($request, 'children', false) ?? false;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php]: same as above

File: src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php [hunk 3]
```diff
@@ -393,7 +393,7 @@ protected function getData(Request $request)
     protected function saveEntity($id, Request $request)
     {
         $parent = $request->get('parent');
-        $breadcrumb = $this->getBooleanRequestParameter($request, 'breadcrumb', false, false);
+        $breadcrumb = $this->getBooleanRequestParameter($request, 'breadcrumb', false) ?? false;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php]: same as above

File: src/Sulu/Bundle/MediaBundle/Entity/Collection.php [hunk 4]
```diff
@@ -88,12 +88,12 @@ class Collection implements CollectionInterface, PermissionInheritanceInterface
     private $key;
 
     /**
-     * @var DoctrineCollection<int, CollectionMeta>
+     * @var DoctrineCollection<array-key, CollectionMeta>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Entity/Collection.php]: this are always `int` as it is not indexed by something else in the mapping. in case all our collections should be not indexed by something else as that would be a performance issue.

File: src/Sulu/Bundle/MediaBundle/Entity/Collection.php [hunk 5]
```diff
@@ -88,12 +88,12 @@ class Collection implements CollectionInterface, PermissionInheritanceInterface
...
      */
     private $meta;
 
     /**
-     * @var DoctrineCollection<int, MediaInterface>
+     * @var DoctrineCollection<array-key, MediaInterface>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Entity/Collection.php]: same as above

File: src/Sulu/Bundle/RouteBundle/Controller/RouteController.php [hunk 6]
```diff
@@ -108,7 +108,7 @@ public function cgetAction(Request $request)
         $id = $this->getRequestParameter($request, 'id', true);
 
         // optional parameter
-        $history = $this->getBooleanRequestParameter($request, 'history', false, false);
+        $history = $this->getBooleanRequestParameter($request, 'history', false) ?? false;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/RouteBundle/Controller/RouteController.php]: same this `??` is not required.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php [hunk 7]
```diff
@@ -169,12 +168,14 @@ public function cgetAction(Request $request)
...
 
         $idsString = $request->get('ids');
+        $locale = $this->getLocale($request);
 
         if ($idsString) {
             $ids = \explode(',', $idsString);
             $snippets = $this->snippetRepository->getSnippetsByUuids($ids, $locale, true);
             $total = \count($snippets);
         } else {
+            $locale = $locale ?? 'de_de';
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:178]: this should not be added here.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php [hunk 8]
```diff
@@ -169,12 +168,14 @@ public function cgetAction(Request $request)
         }
 
         $idsString = $request->get('ids');
+        $locale = $this->getLocale($request);
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php]: would keep it above.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php [hunk 9]
```diff
@@ -427,7 +431,7 @@ public function getSecurityContext()
      * @param string $webspace
      * @param NodeInterface[] $references
      * @param string $id
-     * @param string $locale
+     * @param ?string $locale
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php]: think the locale is required here when calling the contentmapper

File: src/Sulu/Bundle/SnippetBundle/Snippet/SnippetRepository.php [hunk 10]
```diff
@@ -77,7 +77,7 @@ public function getReferences($uuid)
      *
      * UUIDs which fail to resolve to a snippet will be ignored.
      *
-     * @param string $locale
+     * @param ?string $locale
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SnippetBundle/Snippet/SnippetRepository.php:80]: the locale is required here. so would not add any `null` here.

File: UPGRADE.md [hunk 11]
```diff
@@ -13,7 +13,7 @@ Sulu now uses Webpack 5 to build the administration interface application. To en
 - `optimize-css-assets-webpack-plugin` was removed replaced by `css-minimizer-webpack-plugin`: `^4.2.2`
 - `clean-webpack-plugin` was removed and replaced by `clean: true` webpack output option
 - `webpack-clean-obsolete-chunks` was removed and replaced by `clean: true` webpack output option
-- `is-email` was removed and replaced by `sulu-admin-bundle/utils/Email/validateEmail` method 
+- `is-email` was removed and replaced by `sulu-admin-bundle/utils/Email/validateEmail` method
```
Comment (alexander-schranz) [inline UPGRADE.md]: please revert the unnecessary changes to keep here the git histories correctly of this lines

File: src/Sulu/Component/Content/Mapper/ContentMapper.php [hunk 12]
```diff
@@ -292,6 +292,11 @@ public function loadByParent(
         return $children;
     }
 
+    /**
+     * @param string $uuid Path or UUID
+     * @param string|null $locale Locale
```
Comment (alexander-schranz) [inline src/Sulu/Component/Content/Mapper/ContentMapper.php]: locale should stay required here

File: src/Sulu/Component/Rest/RequestParametersTrait.php [hunk 13]
```diff
@@ -25,14 +25,18 @@ trait RequestParametersTrait
      *
      * @param string $name
      * @param bool $force TRUE if value is mandatory
-     * @param mixed $default value if parameter not exists
+     * @param mixed $default value if parameter not exists, this will be ignored if $force is set to true
```
Comment (alexander-schranz) [inline src/Sulu/Component/Rest/RequestParametersTrait.php]: would currently keep it as it was, this would more make problems and could crash existing applications.

File: src/Sulu/Component/Rest/RequestParametersTrait.php [hunk 14]
```diff
@@ -25,14 +25,18 @@ trait RequestParametersTrait
...
      *
      * @return string
      *
      * @throws MissingParameterException parameter is mandatory but does not exists
      */
     protected function getRequestParameter(Request $request, $name, $force = false, $default = null)
     {
+        if ($force) {
+            $default = null;
+        }
```
Comment (alexander-schranz) [inline src/Sulu/Component/Rest/RequestParametersTrait.php]: would not add this, current state I see it more that it could crash existing applications if we change this one.

File: src/Sulu/Component/Rest/RequestParametersTrait.php [hunk 15]
```diff
@@ -49,7 +53,7 @@ protected function getRequestParameter(Request $request, $name, $force = false,
      * @param bool $force TRUE if value is mandatory
      * @param bool $default value if parameter not exists
      *
-     * @return bool
+     * @return bool|null
```
Comment (alexander-schranz) [inline src/Sulu/Component/Rest/RequestParametersTrait.php]: This could be improved to let phpstan if default is bool it always return a bool:

```
	/**
     * @template T of bool|null
     *
	 * @param T $default
     *
	 * @return bool|T
	 */
```

https://phpstan.org/r/01d59854-daa7-4cb7-b003-8fe8ac17101c

**Review by alexander-schranz — CHANGES_REQUESTED** [review]

File: src/Sulu/Component/Rest/RequestParametersTrait.php (hunk 15 above)
Comment (mamazu) [reply src/Sulu/Component/Rest/RequestParametersTrait.php]: Yeah that looks better.

File: src/Sulu/Component/Rest/RequestParametersTrait.php (hunk 14 above)
Comment (mamazu) [reply src/Sulu/Component/Rest/RequestParametersTrait.php]: Removed it.

File: UPGRADE.md (hunk 11 above)
Comment (mamazu) [reply UPGRADE.md]: Removed the BC break so that changing the `UPGRADE.md` wasn't necessary anymore.

File: src/Sulu/Bundle/MediaBundle/Entity/Collection.php (hunk 4 above)
Comment (mamazu) [reply src/Sulu/Bundle/MediaBundle/Entity/Collection.php]: Yes, by default doctrine collections are indexed by ints but in general it could be anything (according to doctrine). Reverting it back to ints upsets the phpstan again, since it can be `array-key` which is just an alias to `string|int`.

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Entity/Collection.php]: yeah but would go with what it really is `int` here and not let phpstan decide what it isnt ;)

**Comment by mamazu** [comment]
Should be fixed.

File: src/Sulu/Bundle/MediaBundle/Entity/Collection.php (hunk 4 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Entity/Collection.php]:
```suggestion
     * @var DoctrineCollection<int, CollectionMeta>
```

Still go with int here, phpstan can currently not detect it, but it is always an int.

File: src/Sulu/Bundle/MediaBundle/Entity/Collection.php (hunk 5 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Entity/Collection.php]:
```suggestion
     * @var DoctrineCollection<int, MediaInterface>
```

Still go with int here, phpstan can currently not detect it, but it is always an int.

File: src/Sulu/Bundle/RouteBundle/Controller/RouteController.php (hunk 6 above)
Comment (alexander-schranz) [inline src/Sulu/Bundle/RouteBundle/Controller/RouteController.php]:
```suggestion
        $history = $this->getBooleanRequestParameter($request, 'history', false);
```

Should not be required.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php (hunk 8 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php]: still open

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php (hunk 7 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:178]: still open this should never be the case that we set here locale to hardcoded value. just remove this line at current state.

```suggestion
```

File: src/Sulu/Bundle/SnippetBundle/Snippet/SnippetRepository.php (hunk 10 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Snippet/SnippetRepository.php:80]: still open, we should keep this also when currently the locale is not handled on all cases correctly:

```suggestion
     * @param string $locale
```

File: src/Sulu/Component/Content/Mapper/ContentMapper.php (hunk 12 above)
Comment (alexander-schranz) [reply src/Sulu/Component/Content/Mapper/ContentMapper.php]: would keep the locale here required and not open the interface also when not handle currently in all cases correctly:

```suggestion
     * @param string $locale
```

File: src/Sulu/Component/Content/Mapper/ContentMapper.php [hunk 16]
```diff
@@ -292,6 +292,11 @@ public function loadByParent(
         return $children;
     }
 
+    /**
+     * @param string $uuid Path or UUID
```
Comment (alexander-schranz) [inline src/Sulu/Component/Content/Mapper/ContentMapper.php]: while looking again on this. this phpdoc should be removed as the phpdoc is defined on the `ContentMapperInterface` and as said the locale should be kept here required. would not open the interface also when not handled currently correclty in all cases.

**Review by alexander-schranz — APPROVED** [review]

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php [hunk 17]
```diff
@@ -274,7 +273,7 @@ public function putAction(Request $request, $id)
      */
     public function deleteAction(Request $request, $id)
     {
-        $locale = $this->getLocale($request);
+        $locale = $this->getLocale($request) ?? 'de_de';
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: we should also not default here to anything hardcoded

File: src/Sulu/Component/Content/ContentTypeManager.php [hunk 18]
```diff
@@ -60,6 +60,11 @@ public function get($alias)
         return $this->container->get($serviceId);
     }
 
+    /**
+     * @param string $alias
+     *
+     * @return bool
+     */
```
Comment (alexander-schranz) [inline src/Sulu/Component/Content/ContentTypeManager.php]: This is defined on the interface and so should not be required to be defined here: https://github.com/sulu/sulu/blob/08a40b7de9c1b3f9a6c42edc928ee8e4ff9ea97c/src/Sulu/Component/Content/ContentTypeManagerInterface.php#L33

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php (hunk 17 above)
Comment (mamazu) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: Well the getLocale function shouldn't return null or else it will cause other problemes. But for now we'll just need to expand the baseline for those errors. 'Cause we can't have function to return something nullable but the places where the value is used not accept null. Would that work for you?

File: src/Sulu/Component/Content/Mapper/ContentMapper.php (hunk 16 above)
Comment (mamazu) [reply src/Sulu/Component/Content/Mapper/ContentMapper.php]: Removed it.

File: src/Sulu/Component/Content/ContentTypeManager.php (hunk 18 above)
Comment (mamazu) [reply src/Sulu/Component/Content/ContentTypeManager.php]: Removed it and fixed the parameter name to match the one documented in the interface.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php (hunk 17 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: Need to check the Permission Request Listener. If we could fallback to `$request->getLocale` in `$this->getLocale()` or if we add a `bool $force` to `getLocale()` method for the cases where it is required.

Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: Okay checked it I would go with `getLocale(Request $request, bool $force = false)` and go with `@return ($force is true ? string : string|null)`

Comment (mamazu) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: I don't like the adding a `$force` parameter idea, since you basically completely changes the function.
 
What should happen when the locale is empty? Throw an exception?

Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: > I don't like the adding a $force parameter idea, since you basically completely changes the function.

The `$force = false` is required as if not forcing the `getLocale` is required to return `null` for the SecurityListener. With the  `bool $force = false` we make sure the method behave the same as befor and only call with with `, true` where we 100% require a locale.

The exception should be this: https://github.com/sulu/sulu/blob/08a40b7de9c1b3f9a6c42edc928ee8e4ff9ea97c/src/Sulu/Component/Rest/RequestParametersTrait.php#L38

Comment (mamazu) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: Makes sense. How about we make a new function for that and deprecate the old one? So that we can have a clean cut between the old and the new behaviour.

Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Controller/SnippetController.php:278]: I need to think about that. At current state I would concentrate in this PR on the PHPDoc and keep the other code as it is so we have code changes itself then in own pull request make searching for issues later easier.

File: src/Sulu/Bundle/AdminBundle/Admin/AdminPool.php [hunk 19]
```diff
@@ -58,6 +58,11 @@ public function getSecurityContexts()
         return $contexts;
     }
 
+    /**
+     * Returns the combined security contexts with placeholders.
+     *
+     * @return array
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Admin/AdminPool.php]: Would like to move this into a seperate pull request where we maybe target the `getSecurityContextsWithPlaceholder()` and `getSecurityContexts` correctly on the `Admin` and `AdminPool` class.

The structure is nearly always the same for getSecurityContexts and so should be for getSecurityContextsWithPlaceholder

System (string) -> Section (string) -> Context (string) -> Permissions (string).

Which ends in:

```php
array<string, array<string, array<string, string[]>>>
```

Comment (mamazu) [reply src/Sulu/Bundle/AdminBundle/Admin/AdminPool.php]: Wouldn't it maybe also make sense to refactor out an interface as the `Admin` and the `AdminPool` are using the same methods.

Comment (mamazu) [reply src/Sulu/Bundle/AdminBundle/Admin/AdminPool.php]: I think we should split this merge request up.

Comment (mamazu) [reply src/Sulu/Bundle/AdminBundle/Admin/AdminPool.php]: Has been split up in #7181 and will be removed from here.

File: src/Sulu/Bundle/AdminBundle/Admin/View/ListItemAction.php [hunk 20]
```diff
@@ -23,7 +23,7 @@ class ListItemAction
     private $type;
 
     /**
-     * @var array
+     * @var array<string, string>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Admin/View/ListItemAction.php]: I'm not sure if this can be even mixed as the JS components decide the options and it could there be also nested things.

```suggestion
     * @var array<string, mixed>
```

File: src/Sulu/Bundle/AdminBundle/Admin/View/ListItemAction.php [hunk 21]
```diff
@@ -35,11 +35,17 @@ public function __construct(string $type, array $options = [])
...
+    /**
+     * @return string
+     */
     public function getType()
     {
         return $this->type;
     }
 
+    /**
+     * @return array<string, string>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Admin/View/ListItemAction.php]:
```suggestion
     * @return array<string, mixed>
```

**Review by alexander-schranz — CHANGES_REQUESTED** [review]
Need to check what we do with ther other comments but think here 2 types we need to change.

**Comment by mamazu** [comment]
This merge request was split up into its parts. That way we can address the lowest version possible for all changes and makes this MR obsolete.
